BZ#1858125 - Include network policy rule optimization (#27168)
openshift/openshift-docs: modules/nw-networkpolicy-optimize.adoc

// Module included in the following assemblies:
//
// * networking/network_policy/about-network-policy.adoc

[id="nw-networkpolicy-optimize_{context}"]
= Optimizations for network policy

> A network policy allows you to isolate pods that are differentiated from one another by labels within a namespace.

Reviewer: ```suggestion
Use a network policy to isolate pods that are differentiated from one another by labels within a namespace.
```

[NOTE]
====
The guidelines for efficient use of network policy rules applies to only the OpenShift SDN cluster network provider.
====

> It is inefficient to apply `NetworkPolicy` objects to large numbers of individual pods in a single namespace. Pod labels do not exist at the IP address level, so network policy generates a separate Open vSwitch (OVS) flow rule for every single possible link between every pod selected with a `podSelector`.

Reviewer: ```suggestion
It is inefficient to apply `NetworkPolicy` objects to large numbers of individual pods in a single namespace. Pod labels do not exist at the IP address level, so a network policy generates a separate Open vSwitch (OVS) flow rule for every possible link between every pod selected with a `podSelector`.
```

For example, if the spec `podSelector` and the ingress `podSelector` within a `NetworkPolicy` object each match 200 pods, then 40,000 (200*200) OVS flow rules are generated. This might slow down a node.

When designing your network policy, refer to the following guidelines:

* Reduce the number of OVS flow rules by using namespaces to contain groups of pods that need to be isolated.

> +

Reviewer 2: (looks like you accidentally pasted a `+` in there)

Author: @danwinship, oops, thanks!

> `NetworkPolicy` objects that select a whole namespace, by using the `namespaceSelector` or an empty `podSelector`, only generate a single OVS flow rule that matches the VXLAN virtual network ID (VNID) of the namespace.

Reviewer: ```suggestion
`NetworkPolicy` objects that select a whole namespace, by using the `namespaceSelector` or an empty `podSelector`, generate only a single OVS flow rule that matches the VXLAN virtual network ID (VNID) of the namespace.
```
